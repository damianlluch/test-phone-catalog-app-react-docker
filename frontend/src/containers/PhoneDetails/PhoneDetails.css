.PhoneDetails {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  margin: 20px 0 50px 0;
}

.PhoneDetails article {
  background-color: #FFF;
  color: #5F6368;
  margin: 10px;
  width: 720px;
  max-width: 720px;
  padding: 15px;
  border-radius: 15px;
}

.PhoneDetails article img {
  height: 300px;
}

.PhoneDetails article p {
  padding: 3px;
}

.PhoneDetails article h1 {
  color: green;
}

@media (max-width: 1000px) {
  .PhoneDetails article {
    width: 70%;    
    padding: 8px;
    margin: 8px;  
  }
}

@media (max-width: 600px) {
  .PhoneDetails article {
    width: 80%;
    padding: 10px;
    margin: 6px;    
  }
}
