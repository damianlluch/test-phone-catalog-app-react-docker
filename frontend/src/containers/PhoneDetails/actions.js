import axios from 'axios';
import C from '../../constants';

export const FETCH = '@Phone/FETCH';
export const fetch = () => ({
  type: FETCH,
});

export const FETCH_SUCCESS = '@Phone/FETCH_SUCCESS';
export const fetchSuccess = phone => ({
  type: FETCH_SUCCESS,
  payload: { phone },
});

export const FETCH_FAILURE = '@Phone/FETCH_FAILURE';
export const fetchFailure = error => ({
  type: FETCH_FAILURE,
  payload: { error },
});

export const RESET = '@Phone/RESET';
export const reset = () => ({
  type: RESET,
});

export const fetchPhone = id => dispatch => {
  dispatch(fetch());
  return axios
    .get(`${C.API_URL}/phones/${id}`)
    .then(res => dispatch(fetchSuccess(res.data)))
    .catch(error => dispatch(fetchFailure(error)));
};
