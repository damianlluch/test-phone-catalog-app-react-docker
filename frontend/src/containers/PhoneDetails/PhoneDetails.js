import React from 'react';
import PropTypes from 'prop-types';
import Spinner from '../../components/Spinner';
import Error from '../../components/Error';
import './PhoneDetails.css';

class PhoneDetails extends React.Component {
  componentDidMount() {
    this.props.fetchPhone(this.props.match.params.id);
  }

  componentWillUnmount() {
    this.props.reset();
  }

  render() {
    const { item, loading, error } = this.props;

    if (loading) {
      return <Spinner />;
    }

    if (error) {
      return <Error error={error.message} />;
    }

    if (item.id) {
      return (
        <section className="PhoneDetails">
          <article>
            <h2>{item.details.title}</h2>
            <br />
            <p><img src={item.details.image} alt={item.details.title} /></p>
          </article>
          <article>
            <h3>Details</h3>
            <br />
            <p><b>Brand</b>: {item.details.brand}</p>
            <p><b>Unveiled</b>: {item.details.unveiled}</p>
            <p><b>Status</b>: {item.details.status}</p>
          </article>
          <article>
            <h3>Design</h3>
            <br />
            <p><b>Size</b>: {item.design.size}</p>
            <p><b>Weight</b>: {item.design.weight}</p>
            <p><b>UsableSurface</b>: {item.design.usableSurface}</p>
            <p><b>Colors</b>: {item.design.colors.join(' - ')}</p>
          </article>
          <article>
            <h3>Screen</h3>
            <br />
            <p><b>Diagonal</b>: {item.screen.diagonal}</p>
            <p><b>AspectRatio</b>: {item.screen.aspectRatio}</p>
            <p><b>Type</b>: {item.screen.type}</p>
            <p><b>Resolution</b>: {item.screen.resolution}</p>
          </article>
          <article>
            <h3>Hardware</h3>
            <br />
            <p><b>CPU</b>: {item.hardware.cpu}</p>
            <p><b>GPU</b>: {item.hardware.gpu}</p>
            <p><b>RAM</b>: {item.hardware.ram}</p>
            <p><b>Capacity</b>: {item.hardware.capacity.join(' - ')}</p>
          </article>
          <article>
            <h3>Antutu Score</h3>
            <br />
            <h1 className="score">{item.performance.antutuScore}</h1>
          </article>
        </section>
      );
    }

    return null;
  }
}

PhoneDetails.propTypes = {
  item: PropTypes.object,
  loading: PropTypes.bool,
  error: PropTypes.object,
  fetchPhone: PropTypes.func,
  match: PropTypes.object,
  reset: PropTypes.object,
};

export default PhoneDetails;
