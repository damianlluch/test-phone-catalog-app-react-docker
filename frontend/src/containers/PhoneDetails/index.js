import { connect } from 'react-redux';
import PhoneDetails from './PhoneDetails';
import { fetchPhone, reset } from './actions';

const mapStateToProps = state => ({
  item: state.phone.item,
  loading: state.phone.loading,
  error: state.phone.error,
});

const mapDispatchToProps = dispatch => ({
  fetchPhone: id => dispatch(fetchPhone(id)),
  reset: () => reset(),
});

export default connect(mapStateToProps, mapDispatchToProps)(PhoneDetails);
