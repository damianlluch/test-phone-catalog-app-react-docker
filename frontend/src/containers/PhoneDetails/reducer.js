import { FETCH, FETCH_SUCCESS, FETCH_FAILURE, RESET } from './actions';

const INITIAL_STATE = {
  item: {},
  loading: false,
  error: null,
};

export default (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case FETCH:
      return {
        ...state,
        loading: true,
        error: null,
      };

    case FETCH_SUCCESS:
      return {
        ...state,
        item: action.payload.phone,
        loading: false,
      };

    case FETCH_FAILURE:
      return {
        ...state,
        item: {},
        loading: false,
        error: action.payload.error,
      };

    case RESET:
      return {
        ...state,
        loading: true,
        error: null,
      };

    default:
      return state;
  }
};
