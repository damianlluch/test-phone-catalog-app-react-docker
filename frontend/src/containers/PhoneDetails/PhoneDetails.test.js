import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import PhoneDetails from './PhoneDetails';

const phone = {
  id: 'iphone',
  details: {
    title: 'Apple iPhone X',
    brand: 'Apple',
    image:
      'https://d2giyh01gjb6fi.cloudfront.net/phone_front/0001/88/thumb_87957_phone_front_big.jpeg',
    unveiled: '2017-09',
    status: 'On Sale',
  },
  design: {
    size: '70.9 mm x 143.6 mm x 7.7 mm',
    weight: '174 g',
    usableSurface: '81 %',
    colors: ['Black', 'Cyan', 'Gray', 'Pink'],
  },
  screen: {
    diagonal: '5.8',
    aspectRatio: '19:9',
    type: 'Oled',
    resolution: '1125 x 2436 px',
  },
  hardware: {
    cpu: 'Apple A11 Bionic',
    gpu: 'Apple three-core graphics processing unit + M11 mo',
    ram: '3 GB',
    capacity: ['64 GB', '256 GB'],
  },
  performance: { antutuScore: '245.149' },
};

describe('<PhoneDetails/>', () => {
  it('renders by default', () => {
    const tree = renderer.create(<MemoryRouter><PhoneDetails match={{ params: { id: 'iphone' } }} fetchPhone={() => { }} item={phone} /></MemoryRouter>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders an error', () => {
    const tree = renderer.create(<PhoneDetails match={{ params: { id: 'iphone' } }} fetchPhone={() => { }} error={{}} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
