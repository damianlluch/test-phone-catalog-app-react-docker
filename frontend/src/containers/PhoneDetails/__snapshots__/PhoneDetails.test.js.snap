// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PhoneDetails/> renders an error 1`] = `
<div
  className="Error"
>
  <h2 />
</div>
`;

exports[`<PhoneDetails/> renders by default 1`] = `
<section
  className="PhoneDetails"
>
  <article>
    <h2>
      Apple iPhone X
    </h2>
    <br />
    <p>
      <img
        alt="Apple iPhone X"
        src="https://d2giyh01gjb6fi.cloudfront.net/phone_front/0001/88/thumb_87957_phone_front_big.jpeg"
      />
    </p>
  </article>
  <article>
    <h3>
      Details
    </h3>
    <br />
    <p>
      <b>
        Brand
      </b>
      : 
      Apple
    </p>
    <p>
      <b>
        Unveiled
      </b>
      : 
      2017-09
    </p>
    <p>
      <b>
        Status
      </b>
      : 
      On Sale
    </p>
  </article>
  <article>
    <h3>
      Design
    </h3>
    <br />
    <p>
      <b>
        Size
      </b>
      : 
      70.9 mm x 143.6 mm x 7.7 mm
    </p>
    <p>
      <b>
        Weight
      </b>
      : 
      174 g
    </p>
    <p>
      <b>
        UsableSurface
      </b>
      : 
      81 %
    </p>
    <p>
      <b>
        Colors
      </b>
      : 
      Black - Cyan - Gray - Pink
    </p>
  </article>
  <article>
    <h3>
      Screen
    </h3>
    <br />
    <p>
      <b>
        Diagonal
      </b>
      : 
      5.8
    </p>
    <p>
      <b>
        AspectRatio
      </b>
      : 
      19:9
    </p>
    <p>
      <b>
        Type
      </b>
      : 
      Oled
    </p>
    <p>
      <b>
        Resolution
      </b>
      : 
      1125 x 2436 px
    </p>
  </article>
  <article>
    <h3>
      Hardware
    </h3>
    <br />
    <p>
      <b>
        CPU
      </b>
      : 
      Apple A11 Bionic
    </p>
    <p>
      <b>
        GPU
      </b>
      : 
      Apple three-core graphics processing unit + M11 mo
    </p>
    <p>
      <b>
        RAM
      </b>
      : 
      3 GB
    </p>
    <p>
      <b>
        Capacity
      </b>
      : 
      64 GB - 256 GB
    </p>
  </article>
  <article>
    <h3>
      Antutu Score
    </h3>
    <br />
    <h1
      className="score"
    >
      245.149
    </h1>
  </article>
</section>
`;
